### PR TITLE
[tests] Add disabled NET6 availability attribute test

### PR DESCRIPTION
**Pull request by chamons**
This test is double disabled:

- Completely disabled until mellite is re-run with a hacked platform assembly and default platform assembly
- Once that's fixed, I'll enable the test, but it'll still be disabled on generated code (via HasCodegenAttribute) until the generator is also fixed. 

But I wanted to get it in and reviewed sooner.

### CODE REVIEW

**Review by dalexsoto — APPROVED** [review]

File: tests/cecil-tests/AttributeTest.cs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+		// In this example, Extension is _not_ considered part of maccatalyst1.0
+		// Because having _any_ SupportedOSPlatform implies only the set explicitly set on that member
+		//
+		// This test should find Extension, note that it has an ios attribute,
+		// and insist that some maccatalyst must also be set.
+		// [TestCaseSource (typeof (Helper), "NetPlatformAssemblies")]
+		public void ChildElementsListAvailabilityForAllPlatformsOnParent (string assemblyPath)
+		{
+			var assembly = Helper.GetAssembly (assemblyPath);
+			if (assembly == null) {
```
Comment (mandel-macaque) [inline tests/cecil-tests/AttributeTest.cs:40]: can we start using the new c# and use is?

File: tests/cecil-tests/AttributeTest.cs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+			// XXX - For now skip generated code until associated generator.cs changes are in
+			if (HasCodegenAttribute(item)) {
+				return;
+			}
+// #if DEBUG
+// 			const string Filter = "AppKit";
+// 			if (!fullName.Contains (" " + Filter)) {
+// 				return;
+// 			}
+// #endif
```
Comment (mandel-macaque) [inline tests/cecil-tests/AttributeTest.cs:76]: remote it? or uncomment?

File: tests/cecil-tests/AttributeTest.cs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+		
+			var parentAvailability = GetAvailabilityAttributes(parent, includeUnsupported: false).ToList();
+// 			// This is true in theory, but our code should be explicit and list every platform individually
+// 			// This can be re-enabled if that decision is reverted.
+
+// 			// iOS implies maccatalyst, but only for parent scope
+// 			if (parentAvailability.Contains("ios") && !parentAvailability.Contains("maccatalyst")) {
+// 				parentAvailability.Append("maccatalyst");
+// 			}
```
Comment (mandel-macaque) [inline tests/cecil-tests/AttributeTest.cs:85]: same.

File: tests/cecil-tests/AttributeTest.cs [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+			var myAvailability = GetAvailabilityAttributes(item, includeUnsupported: true);
+			if (!FirstContainsAllOfSecond (myAvailability, parentAvailability)) {
+				if (!Ignore (fullName)) {
```
Comment (mandel-macaque) [inline tests/cecil-tests/AttributeTest.cs:89]: moe the Ignore to the top and remove the extra indentation.

File: tests/cecil-tests/AttributeTest.cs [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+		IEnumerable<string> GetAvailabilityAttributes(TypeDefinition definition, bool includeUnsupported) => GetAvailabilityAttributes(definition.CustomAttributes, includeUnsupported);
+		IEnumerable<string> GetAvailabilityAttributes(PropertyDefinition prop, bool includeUnsupported) => GetAvailabilityAttributes(prop.CustomAttributes, includeUnsupported);
+		IEnumerable<string> GetAvailabilityAttributes(MethodDefinition meth, bool includeUnsupported) => GetAvailabilityAttributes(meth.CustomAttributes, includeUnsupported);
+
+		IEnumerable<string> GetAvailabilityAttributes(IEnumerable<CustomAttribute> attributes, bool includeUnsupported)
+		{
+			var availability = new List<string>();
+			foreach (var attribute in attributes.Where(a => IsAvailabilityAttribute(a, includeUnsupported))) {
+				var kind = FindAvailabilityKind (attribute);
+				if (kind != null) {
```
Comment (mandel-macaque) [inline tests/cecil-tests/AttributeTest.cs:142]: is not

**Review by mandel-macaque — APPROVED** [review]
Looks good, not blocking for my nits, please clean them a little :)

File: tests/cecil-tests/AttributeTest.cs (hunk 2 above)
Comment (dalexsoto) [reply tests/cecil-tests/AttributeTest.cs:76]: I think this is fine to keep until we are done with the test in general

File: tests/cecil-tests/AttributeTest.cs (hunk 3 above)
Comment (dalexsoto) [reply tests/cecil-tests/AttributeTest.cs:85]: same as above

File: tests/cecil-tests/AttributeTest.cs (hunk 2 above)
Comment (chamons) [reply tests/cecil-tests/AttributeTest.cs:76]: I'm planning on landing these just to uncomment out when debugging the mel and generator changes.

File: tests/cecil-tests/AttributeTest.cs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+				case MethodDefinition meth:
+					return GetAvailabilityAttributes (meth, includeUnsupported);
+				default:
+					throw new NotImplementedException ();
+			}
+		}
+
+		IEnumerable<string> GetAvailabilityAttributes(TypeDefinition definition, bool includeUnsupported) => GetAvailabilityAttributes(definition.CustomAttributes, includeUnsupported);
+		IEnumerable<string> GetAvailabilityAttributes(PropertyDefinition prop, bool includeUnsupported) => GetAvailabilityAttributes(prop.CustomAttributes, includeUnsupported);
+		IEnumerable<string> GetAvailabilityAttributes(MethodDefinition meth, bool includeUnsupported) => GetAvailabilityAttributes(meth.CustomAttributes, includeUnsupported);
```
Comment (rolfbjarne) [inline tests/cecil-tests/AttributeTest.cs:133]:
```suggestion
		IEnumerable<string> GetAvailabilityAttributes (ICustomAttributeProvider provider, bool includeUnsupported) => GetAvailabilityAttributes (provider.CustomAttributes, includeUnsupported);
```

File: tests/cecil-tests/AttributeTest.cs [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+		bool HasCodegenAttribute(PropertyDefinition prop) => HasCodegenAttribute(prop.CustomAttributes);
+		bool HasCodegenAttribute(MethodDefinition meth) => HasCodegenAttribute(meth.CustomAttributes);
+		bool HasCodegenAttribute(IEnumerable<CustomAttribute> attributes) => attributes.Any(a => a.AttributeType.Name == "BindingImplAttribute");
+
+		string? FindAvailabilityKind (CustomAttribute attribute)
+		{
+			if (attribute.ConstructorArguments.Count == 1 && attribute.ConstructorArguments[0].Type.Name == "String") {
+				string full = (string)attribute.ConstructorArguments[0].Value;
+				switch (full) {
+					case string s when full.StartsWith("ios"):
```
Comment (rolfbjarne) [inline tests/cecil-tests/AttributeTest.cs:168]:
```suggestion
					case string s when full.StartsWith("ios", StringComparison.Ordinal):
```

File: tests/cecil-tests/AttributeTest.cs [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+using Mono.Cecil;
+
+#nullable enable
+
+namespace Cecil.Tests {
+
+	// [TestFixture]
+	public class AttributeTest {
+		// https://github.com/xamarin/xamarin-macios/issues/10170
```
Comment (rolfbjarne) [inline tests/cecil-tests/AttributeTest.cs:16]: Another test scenario:

Every API must either have no attributes at all, or have an availability attribute for the current platform (i.e. every type (or method or property or ...) in Xamarin.iOS.dll must have a `[SupportedOSPlatform ("ios*")]` or no availability attributes at all).

Except: enum fields, for enums we often (on purpose) include fields that only exists on other platforms.

File: tests/cecil-tests/AttributeTest.cs (hunk 6 above)
Comment (chamons) [reply tests/cecil-tests/AttributeTest.cs:133]: mindblown.jpg

**Comment by chamons** [comment]
Added that second suggested test.

File: tests/cecil-tests/AttributeTest.cs [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+		{
+			return provider.CustomAttributes.Any(a => a.AttributeType.Name == "BindingImplAttribute");
+		}
+
+		string? FindAvailabilityKind (CustomAttribute attribute)
+		{
+			if (attribute.ConstructorArguments.Count == 1 && attribute.ConstructorArguments[0].Type.Name == "String") {
+				string full = (string)attribute.ConstructorArguments[0].Value;
+				switch (full) {
+					case string s when full.StartsWith("ios", StringComparison.Ordinal):
```
Comment (tj-devel709) [inline tests/cecil-tests/AttributeTest.cs:208]: Woah, I've never seen this before!

File: tests/cecil-tests/AttributeTest.cs [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+			return IsSupportedAttribute (attribute) ||
+				(includeUnsupported && attribute.AttributeType.Name == "UnsupportedOSPlatformAttribute");
+		}
+
+		bool IsSupportedAttribute (CustomAttribute attribute)
+		{
+			return attribute.AttributeType.Name == "SupportedOSPlatformAttribute";
+		}
+	}
+}
```
Comment (tj-devel709) [inline tests/cecil-tests/AttributeTest.cs:241]: nit: missing newline

**Review by tj-devel709 — APPROVED** [review]

File: tests/cecil-tests/AttributeTest.cs [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+		// {
+		//     public static void Original () { }
+		//
+		//     [SupportedOSPlatform(""ios2.0"")]
+		//     public static void Extension () { }
+		// }
+		//
+		// When run against mac, this fails as Extension does not include a mac supported of any kind attribute
+		// [TestCaseSource (typeof (Helper), "NetPlatformAssemblies")]
+		public void AllAttributedItemsMostIncludeCurrentPlatform (string assemblyPath)
```
Comment (rolfbjarne) [inline tests/cecil-tests/AttributeTest.cs:105]:
```suggestion
		public void AllAttributedItemsMustIncludeCurrentPlatform (string assemblyPath)
```

File: tests/cecil-tests/AttributeTest.cs [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+
+#if DEBUG
+			Console.WriteLine(assemblyPath);
+#endif
+
+			HashSet<string> found = new HashSet<string> ();
+			foreach (var prop in Helper.FilterProperties (assembly, a => HasAnyAvailabilityAttribute (a, includeUnsupported: true))) {
+				CheckAllPlatformsOnParent (prop, prop.FullName, prop.DeclaringType, found);
+			}
+			foreach (var meth in Helper.FilterMethods (assembly, a => HasAnyAvailabilityAttribute (a, includeUnsupported: true))) {
```
Comment (rolfbjarne) [inline tests/cecil-tests/AttributeTest.cs:54]: What about fields? They have availability attributes too.

File: tests/cecil-tests/AttributeTest.cs [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+			case "Xamarin.iOS.dll":
+				return "ios";
+			case "Xamarin.TVOS.dll":
+				return "tvos";
+			case "Xamarin.Mac.dll":
+				return "macos";
+			case "Xamarin.MacCatalyst.dll":
+				return "maccatalyst";
+			default:
+				throw new NotImplementedException ();
```
Comment (rolfbjarne) [inline tests/cecil-tests/AttributeTest.cs:155]: The assembly names will change soon, so implementing it this way makes it not require any changes when that happens:

```suggestion
			var baseName = Path.GetFileName (assemblyPath));
			if (Configuration.GetBaseLibraryName (TargetFramework.DotNet_5_0_iOS) == baseName)
				return "ios";
			if (Configuration.GetBaseLibraryName (TargetFramework.DotNet_5_0_tvOS) == baseName)
				return "tvos";
			if (Configuration.GetBaseLibraryName (TargetFramework.DotNet_5_0_macOS) == baseName)
				return "macos";
			if (Configuration.GetBaseLibraryName (TargetFramework.DotNet_5_0_MacCatalyst) == baseName)
				return "maccatalyst";
			throw new NotImplementedException ();
```

File: tests/cecil-tests/AttributeTest.cs [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+				var kind = FindAvailabilityKind (attribute);
+				if (kind is not null) {
+					availability.Add (kind);
+				}
+			}
+			return availability;
+		}
+
+		// Unfortunate state we need to keep since I can't see to walk "up" from a
+		// MethodDefinition get_Foo or SetFoo to see it's container's properties
```
Comment (rolfbjarne) [inline tests/cecil-tests/AttributeTest.cs:197]: Something like this should work:

```cs
var getter = ...;
var property = getter.DeclaringType.Properties.Find (v => v.Name == getter.Name.Substring (4));
```

Comment (chamons) [reply tests/cecil-tests/AttributeTest.cs:209]: It seems to work, at least enough for this PR. I'll fix it up with the attribute changes in follow up.

**Review by rolfbjarne — APPROVED** [review]

**Pull request merged**